File: backend/src/main/java/com/dasigconnect/backend/service/PasswordPolicy.java
package com.dasigconnect.backend.service;

import java.util.List;
import java.util.Locale;

import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;

final class PasswordPolicy {

    static final int MIN_LENGTH = 12;

    private static final List<String> COMMON_FRAGMENTS = List.of(
            "password",
            "qwerty",
            "admin",
            "dasig",
            "welcome",
            "letmein",
            "123",
            "abc");

    private PasswordPolicy() {
    }

    static void validate(String password, String email, String firstName, String lastName) {
        if (password == null || password.isBlank()) {
            reject("Password is required.");
        }
        if (password.length() < MIN_LENGTH) {
            reject("Password must be at least 12 characters.");
        }
        if (password.chars().anyMatch(Character::isWhitespace)) {
            reject("Password cannot contain spaces.");
        }
        if (password.chars().noneMatch(Character::isUpperCase)) {
            reject("Password must include an uppercase letter.");
        }
        if (password.chars().noneMatch(Character::isLowerCase)) {
            reject("Password must include a lowercase letter.");
        }
        if (password.chars().noneMatch(Character::isDigit)) {
            reject("Password must include a number.");
        }
        if (password.chars().allMatch(Character::isLetterOrDigit)) {
            reject("Password must include a special character.");
        }

        String normalized = password.toLowerCase(Locale.ROOT);
        for (String fragment : COMMON_FRAGMENTS) {
            if (normalized.contains(fragment)) {
                reject("Password is too easy to guess.");
            }
        }
        rejectIfContainsIdentity(normalized, email);
        rejectIfContainsIdentity(normalized, firstName);
        rejectIfContainsIdentity(normalized, lastName);
    }

    private static void rejectIfContainsIdentity(String normalizedPassword, String value) {
        if (value == null || value.isBlank()) {
            return;
        }
        String normalized = value.trim().toLowerCase(Locale.ROOT);
        int at = normalized.indexOf('@');
        if (at > 0) {
            normalized = normalized.substring(0, at);
        }
        normalized = normalized.replaceAll("[^\\p{Alnum}]", "");
        if (normalized.length() >= 3 && normalizedPassword.contains(normalized)) {
            reject("Password cannot contain your name or email.");
        }
    }

    private static void reject(String message) {
        throw new ResponseStatusException(HttpStatus.BAD_REQUEST, message);
    }
}

File: frontend/src/lib/passwordPolicy.ts
export interface PasswordRules {
  length: boolean;
  upper: boolean;
  lower: boolean;
  number: boolean;
  symbol: boolean;
  noSpaces: boolean;
  notCommon: boolean;
  noIdentity: boolean;
}

const COMMON_FRAGMENTS = [
  "password",
  "qwerty",
  "admin",
  "dasig",
  "welcome",
  "letmein",
  "123",
  "abc",
];

export function getPasswordRules(
  password: string,
  identityValues: Array<string | undefined | null> = [],
): PasswordRules {
  const normalized = password.toLowerCase();
  return {
    length: password.length >= 12,
    upper: /[A-Z]/.test(password),
    lower: /[a-z]/.test(password),
    number: /[0-9]/.test(password),
    symbol: /[^A-Za-z0-9]/.test(password),
    noSpaces: !/\s/.test(password),
    notCommon: !COMMON_FRAGMENTS.some((part) => normalized.includes(part)),
    noIdentity: !identityValues.some((value) =>
      containsIdentityFragment(normalized, value),
    ),
  };
}

export function isStrongPassword(
  password: string,
  identityValues: Array<string | undefined | null> = [],
) {
  return Object.values(getPasswordRules(password, identityValues)).every(Boolean);
}

export function firstPasswordError(
  password: string,
  identityValues: Array<string | undefined | null> = [],
) {
  const rules = getPasswordRules(password, identityValues);
  if (!rules.length) return "Password must be at least 12 characters.";
  if (!rules.noSpaces) return "Password cannot contain spaces.";
  if (!rules.upper) return "Password must include an uppercase letter.";
  if (!rules.lower) return "Password must include a lowercase letter.";
  if (!rules.number) return "Password must include a number.";
  if (!rules.symbol) return "Password must include a special character.";
  if (!rules.notCommon) return "Password is too easy to guess.";
  if (!rules.noIdentity) return "Password cannot contain your name or email.";
  return "";
}

function containsIdentityFragment(
  normalizedPassword: string,
  value: string | undefined | null,
) {
  if (!value) return false;
  const localPart = value.trim().toLowerCase().split("@")[0] ?? "";
  const fragments = localPart
    .split(/[^a-z0-9]+/g)
    .map((part) => part.replace(/[^a-z0-9]/g, ""))
    .filter((part) => part.length >= 3);
  const compact = localPart.replace(/[^a-z0-9]/g, "");
  if (compact.length >= 3) {
    fragments.push(compact);
  }
  return fragments.some((part) => normalizedPassword.includes(part));
}
